File: Week 5/python/6.recursion_patterns.py
"""
WEEK 5 - PYTHON FUNCTIONS & RECURSION
Topic: Common Recursion Patterns
File: 6.recursion_patterns.py

CONCEPT:
Mastering a handful of recursion patterns covers ~90% of DSA recursion
problems:
  1. Linear recursion -- one recursive call per invocation (factorial, sum, reverse)
  2. Binary recursion -- two recursive calls (Fibonacci, binary search, merge sort)
  3. Tail recursion -- the recursive call is the LAST operation
  4. Mutual recursion -- function A calls B; B calls A (e.g., isEven/isOdd)
  5. Helper / accumulator -- carry state in extra parameters
  6. Subset / backtracking -- explore each choice (include / exclude)

KEY POINTS:
 - Python doesn't optimise tail calls -- depth still costs stack frames.
 - Closures + mutable captures replace Java's "static counter" idioms.
 - For SUBSETS / PERMUTATIONS, recursion is the standard technique.
"""


# Pattern 1: Linear recursion -- reverse a string
def reverse_str(s: str) -> str:
    if not s:
        return ""
    return reverse_str(s[1:]) + s[0]


# Pattern 3: Tail recursion with accumulator -- factorial
def fact_tail(n: int, acc: int = 1) -> int:
    if n <= 1:
        return acc
    return fact_tail(n - 1, n * acc)


# Pattern 4: Mutual recursion
def is_even(n: int) -> bool:
    if n == 0:
        return True
    return is_odd(n - 1)


def is_odd(n: int) -> bool:
    if n == 0:
        return False
    return is_even(n - 1)


# Pattern 5: Sum of digits via recursion
def digit_sum(n: int) -> int:
    if n == 0:
        return 0
    return (n % 10) + digit_sum(n // 10)


# Pattern 6: Generate all subsets of a list (2^n total)
def subsets(arr: list[int], idx: int = 0, current: list[int] | None = None) -> None:
    if current is None:
        current = []
    if idx == len(arr):
        print("{" + ",".join(str(x) for x in current) + "}")
        return
    # Choice 1: exclude arr[idx]
    subsets(arr, idx + 1, current)
    # Choice 2: include arr[idx]
    subsets(arr, idx + 1, current + [arr[idx]])


# Bonus: palindrome check via recursion
def is_palindrome(s: str, left: int = 0, right: int | None = None) -> bool:
    if right is None:
        right = len(s) - 1
    if left >= right:
        return True
    if s[left] != s[right]:
        return False
    return is_palindrome(s, left + 1, right - 1)


def main() -> None:
    print("reverse('hello') =", reverse_str("hello"))
    print("fact_tail(5)    =", fact_tail(5))
    print("fact_tail(10)   =", fact_tail(10))

    print("is_even(4) =", is_even(4))
    print("is_odd(7)  =", is_odd(7))

    print("digit_sum(1234) =", digit_sum(1234))

    print("\nAll subsets of [1, 2, 3]:")
    subsets([1, 2, 3])

    print("\nis_palindrome('racecar') =", is_palindrome("racecar"))
    print("is_palindrome('hello')   =", is_palindrome("hello"))


if __name__ == "__main__":
    main()


# NOTES:
# - Python list slicing (s[1:]) is O(n) and allocates -- watch the cost in deep recursion.
# - For tail recursion, prefer iterative form because Python doesn't optimise TCO.
# - Subsets via recursion is the foundation of BACKTRACKING (Week 14+ material).
# - Use itertools.combinations / permutations for ready-made enumerations.

File: Week 5/java/6.RecursionPatterns.java
/*
 * WEEK 5 - FUNCTIONS & RECURSION
 * Topic: Common Recursion Patterns
 *
 * Master these patterns and you can solve 90% of recursive problems:
 *
 * PATTERN 1: Linear Recursion — one recursive call per invocation
 *   Example: factorial, sum, reverse string
 *
 * PATTERN 2: Binary Recursion — two recursive calls per invocation
 *   Example: Fibonacci, binary search, merge sort
 *
 * PATTERN 3: Tail Recursion — recursive call is the LAST operation
 *   Advantage: can be optimized by compiler to use O(1) stack space
 *   (Java doesn't optimize this automatically, but it's a good concept)
 *
 * PATTERN 4: Mutual Recursion — function A calls B, B calls A
 *   Example: isEven/isOdd check
 *
 * PATTERN 5: Helper / Accumulator pattern — carry state in parameters
 *
 * Time/Space: varies by problem (shown per example)
 */

public class RecursionPatterns {

    // PATTERN 1: Linear Recursion — Reverse a string
    // Time: O(n), Space: O(n) stack + O(n) output
    static String reverse(String s) {
        if (s.isEmpty()) return "";
        return reverse(s.substring(1)) + s.charAt(0);
    }

    // PATTERN 3: Tail Recursion — Factorial with accumulator
    // tail-recursive because the last operation is the recursive call itself
    static long factTail(int n, long acc) {
        if (n <= 1) return acc;
        return factTail(n - 1, n * acc); // accumulate result into 'acc'
    }
    // Wrapper: start accumulator at 1
    static long factorial(int n) { return factTail(n, 1); }

    // PATTERN 4: Mutual Recursion
    static boolean isEven(int n) {
        if (n == 0) return true;
        return isOdd(n - 1);
    }
    static boolean isOdd(int n) {
        if (n == 0) return false;
        return isEven(n - 1);
    }

    // PATTERN 5: Helper / Accumulator — Sum digits of a number
    static int digitSum(int n) {
        if (n == 0) return 0;
        return (n % 10) + digitSum(n / 10);
    }

    // BONUS: Recursion to generate all subsets (2^n subsets)
    // This is the foundation of backtracking
    static void subsets(int[] arr, int idx, String current) {
        if (idx == arr.length) {
            System.out.println("{" + current + "}");
            return;
        }
        // Choice 1: EXCLUDE arr[idx]
        subsets(arr, idx + 1, current);
        // Choice 2: INCLUDE arr[idx]
        String sep = current.isEmpty() ? "" : ",";
        subsets(arr, idx + 1, current + sep + arr[idx]);
    }

    // BONUS: Check if a string is a palindrome using recursion
    static boolean isPalindrome(String s, int left, int right) {
        if (left >= right) return true;
        if (s.charAt(left) != s.charAt(right)) return false;
        return isPalindrome(s, left + 1, right - 1);
    }

    public static void main(String[] args) {
        // Pattern 1
        System.out.println("reverse(\"hello\") = " + reverse("hello"));

        // Pattern 3
        System.out.println("factorial(5) [tail] = " + factorial(5));
        System.out.println("factorial(10) [tail] = " + factorial(10));

        // Pattern 4
        System.out.println("isEven(4) = " + isEven(4));
        System.out.println("isOdd(7) = " + isOdd(7));

        // Pattern 5
        System.out.println("digitSum(1234) = " + digitSum(1234)); // 10

        // Subsets
        System.out.println("\nAll subsets of {1, 2, 3}:");
        subsets(new int[]{1, 2, 3}, 0, "");

        // Palindrome
        System.out.println("\nisPalindrome(\"racecar\") = " + isPalindrome("racecar", 0, 6));
        System.out.println("isPalindrome(\"hello\") = " + isPalindrome("hello", 0, 4));
    }
}
